### ISSUE TITLE
MsgPack object improvements

### ISSUE THREAD

**Issue by locker** [tarantool/tarantool#6823]
Follow-up #6666 improvements suggested by @Gerold103:

1. Add a function to determine the type of the value under MsgPack iterator cursor or at least a function to check if the value at the iterator cursor is nil. Rationale: sometimes one needs to strip nil from a return value. Example:
   ```lua
   local mp = msgpack.object({123, nil})
   local it = mp:iterator()
   local mp2 = it:take()
   if it:is_nil() then
       return mp2
   else
       return mp
   end
   ```
2. Add a function to return array tail as msgpack object. Rationale: there are cases when we want to drop the first element of an array and forward the tail (e.g. the first element is a status, and the rest is the value). Example:
   ```lua
   local mp = msgpack.object({true, 1, 2, 3})
   local it = mp:iterator()
   if it:decode() then
       return it:take_array_tail()
   else
       return false
   end
   ```

**Comment by olegrok** [comment]
Is it possible to implement msgpack object constuctor?

There is a flatten procedure on a router that form array (tuple) from lua table.
E.g.
```
object = {id = 1, name = "name"}
tuple = {object.id, object.name}
```

Probably it's better to omit temporary table creation and put data directly to msgpack object.
```
data = msgpack.new_object()
data:push(object.id)
data:push(object.name)
vshard.router.call(..., {data})
```

Probably it could improve router performance as well. But it's hard to evaluate possible impact.

**Comment by R-omk** [comment]
> it:take_array_tail()

`Slice(offset int, length int)`

Ex:
Slice(-3) - last 3
Slice(1, -2)  - start from first, omit last two.

**Comment by Gerold103** [comment]
So far I only need `iterator:take()`, `iterator:decode()`, `iterator:take_array_tail()`/`:slice()`. The `iterator:is_nil()` while overall might be useful in future, won't help me much here. Because on the storage I have this:
```Lua
ok, ret1, ret2, ret3 = local_call(name, args)
if not ok then
    ret1 = lerror.make(ret1)
end
return ok, ret1, ret2, ret3
```
The problem here is that even if your function returns just one value (`ret1`), `ret2` and `ret3` are still returned too and are turned into `box.NULL`. On the router I get them in the end. So to skip them I will need to iterate all the results using `iterator:take()`. The result will be an array of msgpack objects, on whose items I won't be able to call `:is_nil()`, so for each of them starting from tail I will need to create an iterator just for that check. Each returned value will cost one msgpack object and one iterator - too much pressure on Lua GC for scalar results.

My plan is to patch the storage to trim the nils on its end and on the router just use `:slice()`/`:take_array_tail()` for whatever comes from the storage. The downside is that even if you actually return 2 or 3 values but the tail is nils, I will truncate it on the storage. It could be fixed with variadic arguments and tail call hacks in Lua to return exactly as many values as needed, but variadic args can't be jitted.

**TLDR**: I only need one addition so far: `iterator:take_array_tail()`/`:slice()`.
